src/libraries/BitMath.sol

```diff
@@ -3,96 +3,47 @@ pragma solidity ^0.8.20;
 
 /// @title BitMath
 /// @dev This library provides functionality for computing bit properties of an unsigned integer
+/// @author Solady (https://github.com/Vectorized/solady/blob/8200a70e8dc2a77ecb074fc2e99a2a0d36547522/src/utils/LibBit.sol)
 library BitMath {
     /// @notice Returns the index of the most significant bit of the number,
     ///     where the least significant bit is at index 0 and the most significant bit is at index 255
-    /// @dev The function satisfies the property:
-    ///     x >= 2**mostSignificantBit(x) and x < 2**(mostSignificantBit(x)+1)
     /// @param x the value for which to compute the most significant bit, must be greater than 0
     /// @return r the index of the most significant bit
     function mostSignificantBit(uint256 x) internal pure returns (uint8 r) {
         require(x > 0);
 
-        unchecked {
-            if (x >= 0x100000000000000000000000000000000) {
-                x >>= 128;
-                r += 128;
-            }
-            if (x >= 0x10000000000000000) {
-                x >>= 64;
-                r += 64;
-            }
-            if (x >= 0x100000000) {
-                x >>= 32;
-                r += 32;
-            }
-            if (x >= 0x10000) {
-                x >>= 16;
-                r += 16;
-            }
-            if (x >= 0x100) {
-                x >>= 8;
-                r += 8;
-            }
-            if (x >= 0x10) {
-                x >>= 4;
-                r += 4;
-            }
-            if (x >= 0x4) {
-                x >>= 2;
-                r += 2;
-            }
-            if (x >= 0x2) r += 1;
+        assembly ("memory-safe") {
+            r := or(shl(8, iszero(x)), shl(7, lt(0xffffffffffffffffffffffffffffffff, x)))
+            r := or(r, shl(6, lt(0xffffffffffffffff, shr(r, x))))
+            r := or(r, shl(5, lt(0xffffffff, shr(r, x))))
+            r := or(r, shl(4, lt(0xffff, shr(r, x))))
+            r := or(r, shl(3, lt(0xff, shr(r, x))))
+            // forgefmt: disable-next-item
+            r := or(r, byte(and(0x1f, shr(shr(r, x), 0x8421084210842108cc6318c6db6d54be)),
+                0x0706060506020504060203020504030106050205030304010505030400000000))
```

Comment:
should this also mention De Bruijn?

Reply:
This is not De Brujin, but rather a custom lookup with numbers found via z3.